File: pchumToWeb.js
// consts

const header = '<html><head>\n \
<meta http-equiv="content-type" content="text/html; charset=UTF-8">\n \
<title>#%TITLE% | Log for %YEAR%-%MONTH%-%DAY%</title>\n \
<link rel="stylesheet" href="https://jingloria.wertercatt.com/lostpawns/logs2/pesterlog-viewer.css">\n \
<meta charset="UTF-8">\n \
</head>\n \
<body cz-shortcut-listen="true">\n \
<h1>#%TITLE% | Log for %YEAR%-%MONTH%-%DAY%</h1>\n \
<div class="memolist">\n \
MEMOS\n \
</div>\n'


// Init File Reader
var fileReader = new FileReader();
fileReader.onload = event => {
  console.log(logName)
  formatPesterlog(event.target.result)
}

//Load file
var logName = ""
let loadFile = () => {
  var fileToLoad = document.getElementById("input").files[0]
  fileReader.readAsText(fileToLoad, "UTF-8"); 
  logName = fileToLoad.name
}

// Format HTML
let formatPesterlog = plog => {
  var outputDiv = document.getElementById("output")
  var tempformat = document.getElementById("formatter")

  // Get date and time from file name
  var fileInfo = {
    mins: logName.substring(logName.length - 6, logName.length - 4),
    hour: logName.substring(logName.length - 9, logName.length - 7),
    day: logName.substring(logName.length - 12, logName.length - 10),
    month: logName.substring(logName.length - 15, logName.length - 13),
    year: logName.substring(logName.length - 20, logName.length - 16),
    title: logName.substring(0, logName.length - 21)
  }

  // Create header
  var logHeader = header
  for (const [key, value] of Object.entries(fileInfo)) {
    logHeader = logHeader.replaceAll("%" + key.toUpperCase() + "%", value)
  }

  // Create head
  console.log(logHeader)

  // Get user list
  tempformat.innerHTML = plog
  var messageSpans = document.querySelectorAll("#formatter > span")
  var userList = []
  var userInfo = []

  // Get all Abbreviated chumhandles
  messageSpans.forEach(span => {
    if (span.innerText[3] == ":") {
      var chum = span.innerText.substring(1, 3)
      if (!userList.includes(chum)) {
        userInfo.push({
          chum: chum,
          color: span.style.color,
          tense: span.innerText[0]
        })
        userList.push(chum)
      }
    }
  })

  // Format userdivs
  var particiDiv = document.createElement("div")
  particiDiv.classList = "participants"
  userInfo.forEach(obj => {

    // Look for if full chumhandles exist
    var notiPos = tempformat.innerHTML.indexOf("[" + obj.tense + obj.chum + "]")
    if (notiPos !== -1) {
      obj.handle = tempformat.innerHTML.substring(tempformat.innerHTML.lastIndexOf(" ", notiPos - 2) + 1, notiPos - 1)
    }

    // Create userSpans
    var userSpan = document.createElement("span")
    userSpan.style.color = obj.color
    userSpan.innerText = obj.tense + " " + (obj.handle ? obj.handle : "??????") + " [" + obj.tense + obj.chum + "]"
    particiDiv.append(userSpan)
    particiDiv.append(document.createElement("br"))
  })

  console.log(userList, userInfo)

  // TODO: Replace MEMOO with memo div

  var userOuput =  particiDiv.outerHTML.replaceAll("br>", "br>\n").replaceAll('pants">', 'pants">\n')
  var output = logHeader + userOuput + " <div class=memolog>" + plog + "</div></body></html>"


  outputDiv.innerText = output

  download(fileInfo.title + fileInfo.year + "-" + fileInfo.month + "-" + fileInfo.day + ".html", output)
}


// Stolen download function

function download(filename, text) {
  var element = document.createElement('a');
  element.setAttribute('href', 'data:text/plain;charset=utf-8,' + encodeURIComponent(text));
  element.setAttribute('download', filename);

  element.style.display = 'none';
  document.body.appendChild(element);

  element.click();

  document.body.removeChild(element);
}

File: pchumToWeb.py
import os
import sys
from bs4 import BeautifulSoup
import json
import re
from tqdm import tqdm

def convert(filePath):
    header = '<html><head>\n \
    <meta http-equiv="content-type" content="text/html; charset=UTF-8">\n \
    <title>{title} | Log for {year}-{month}-{day} starting at {hour}:{mins}</title>\n \
    <style>body {{ background-color: #101535; font-family: Courier New,Courier,monospace; font-weight: bold; }} h1{{ text-align: center; color: #FFFFFF; }} .memolog {{ text-align: left; background-color: #FFFFFF; border: 5px solid #3F4472; height: 500px; width: 48%; padding: 5px; margin-left: auto; overflow: scroll; margin-right: auto; float: center; }} .participants {{ text-align: left; background-color: #FFFFFF; border: 5px solid #3F4472; height: 500px; width: 22%; padding: 5px; margin-left: auto; overflow: scroll; margin-right: auto; float: right; white-space: nowrap }} .memolist {{ text-align: left; background-color: #FFFFFF; border: 5px solid #3F4472; height: 500px; width: 22%; padding: 5px; margin-left: auto; overflow: scroll; margin-right: auto; white-space: nowrap; float: left; }}</style>\n \
    <meta charset="UTF-8">\n \
    </head>\n \
    <body cz-shortcut-listen="true">\n \
    <h1>{title} | Log for {year}-{month}-{day} starting at {hour}:{mins}</h1>\n \
    <div class="memolist">\n \
    MEMOS\n \
    </div>\n'
    with open(filePath, "r", encoding="utf-8", errors="backslashreplace") as unformattedLog:
        formattedLog = "<!-- Original Filepath: " + filePath.split("Pesterchum-logs")[1] + "-->"
        unformattedLogContents = unformattedLog.read()
        logName = os.path.basename(filePath)
        try:
            mins = logName.split(".")[3]
        except IndexError as e:
            mins = "Unknown"
        try:
            hour = logName.split(".")[2]
        except IndexError as e:
            hour = "Unknown"
        try:
            day = logName.split(".")[1].split("-")[2]
        except IndexError as e:
            day = "Unknown"
        try:
            month = logName.split(".")[1].split("-")[1]
        except IndexError as e:
            month = "Unknown"
        try:
            year = logName.split(".")[1].split("-")[0]
        except IndexError as e:
            year = "Unknown"
        try:
            title = logName.split(".")[0]
        except IndexError as e:
            title = "Unknown"
        logInfo = {
            "mins": mins,
            "hour": hour,
            "day": day,
            "month": month,
            "year": year,
            "title": title,
        }
        logHeader = header.format(**logInfo)
        formattedLog = formattedLog + "\n" + logHeader
        userInfo = []
        unformattedLogParser = BeautifulSoup(unformattedLogContents, 'html.parser')
        messageSpans = unformattedLogParser.find_all("span")
        for span in messageSpans:
            spanText = span.get_text()
            if len(spanText) > 3:
                if spanText[3] == ":":
                    try:
                        color = span["style"]
                    except KeyError as e:
                        color = '"color:#646464"'
                    chum = {
                            "chum": spanText[1:3],
                            "color": color,
                            "tense": spanText[0],
                        }
                    if chum not in userInfo:
                        userInfo.append(chum)
        participantsDiv = '<div class="participants">'
        for user in userInfo:
            userSpan = '<span style=' + user["color"] + ">"
            searchString = re.escape("[" + user["tense"] + user["chum"] + "]")
            searchResult = unformattedLogParser.find(string=re.compile(searchString))
            if searchResult != None:
                userSpan = userSpan + searchResult.get_text() + "</span></br>"
            else:
                userSpan = userSpan + user["tense"] + " ?????? [" + user["tense"] + user["chum"] + "]</span></br>"
            participantsDiv = participantsDiv + "\n" + userSpan
        participantsDiv = participantsDiv + "</div>"
        formattedLog = formattedLog + "\n" + participantsDiv
        formattedLog = formattedLog + "\n<div class=memolog>\n" + unformattedLogContents
        formattedLog = formattedLog + "\n</div>\n</body>\n</html>"
        formattedLogDirectory = filePath.split("Pesterchum-logs")[0] + "Pesterchum-logs\\PrettifiedAndSortedByDate\\" + logInfo["year"] + "-" + logInfo["month"] + "-" + logInfo["day"] + "\\"
        os.makedirs(formattedLogDirectory, exist_ok=True)
        formattedLogFilename = formattedLogDirectory + os.path.basename(filePath)
        while os.path.exists(formattedLogFilename):
            formattedLogFilename = formattedLogFilename + ".dupe.html"
        with open(formattedLogFilename, "w", encoding="utf-8", errors="backslashreplace") as output:
            output.write(formattedLog)
        # print(json.dumps(logHeader, sort_keys=True, indent=4))
        # print(json.dumps(userInfo, sort_keys=True, indent=4))

def bulkConvert(rootPath):
    file_paths = []

    for root, dirs, files in os.walk(rootPath):
        for name in files:
            file_path = os.path.join(root, name)
            if os.path.splitext(file_path)[1] == '.html':
                file_paths += [file_path]
    for log in tqdm(file_paths):
        convert(log)


if __name__ == "__main__":
    bulkConvert(sys.argv[1])